### ISSUE TITLE
Disable code coverage report on Gitlabci

### ISSUE THREAD

**Issue by LaSylv** [mc1arke/sonarqube-community-branch-plugin#285]
**Describe the bug**

Disable the coverage pushed to gitlab, as it is sometimes not correct and gitlab already provides this feature.

**Expected behavior**

It wold be nice to be able to disable the coverage pushed by the plugin through a specific property.

**Additional context**

As we already use the coverage through gitlab (more complete, showing which lines are covered or not), the sonar gitlab is a pain to set and never seem to be correct. We would like to disable it entirely when it's pushed to gitlab.

If this is somzthing you would consider, I can work on a PR

**Comment by lucasoares** [comment]
@Glosur, what you mean about disabling coverage pushed to gitlab?

Is the coverage provided by the external job created by the plugin?

![image](https://user-images.githubusercontent.com/10624972/108893148-82fa3000-75ef-11eb-92ce-75569be0573c.png)

If so, I agree with you because the final coverage displayed on the MR page will be incorrect since it will show the average coverage between my test and the coverage of the new code.

For example, if I have one MR with only 1 changed line that doesn't is covered by tests, the widget will report an incorrect coverage of `(test coverage + sonar coverage) / 2`:

![image](https://user-images.githubusercontent.com/10624972/108892883-2ac32e00-75ef-11eb-8c36-c7c6ee85c907.png)

**Comment by Glosur** [comment]
Yes, we are having the problem you're saying, that is heavily linked with https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/225


We use another gitlabci job to get the covergae from. Gitlab then also receives the coverage from the sonar plugin (the coverage is not the correct one, see PR linked)

In the end, even if the PR is fixed, I have to maintain two tools to have the coverage in the gitlabci job matching the one in sonar

**Comment by lucasoares** [comment]
Even with #225 merged we are having this problem as well.

I just ran an analysis with sonarqube 9.5 and with plugin version 1.12.0 and the reported coverage in the SonarQube external job is the new code coverage and not the total coverage.

![image](https://user-images.githubusercontent.com/10624972/184224425-40e446b9-2a94-4803-a91d-c96dbd709642.png)

**Comment by Karql** [comment]
Hi @mc1arke 

Regression was made in this change: https://github.com/mc1arke/sonarqube-community-branch-plugin/commit/ce928934dfaa11b9edb7933970b3a7373bd6b151#diff-22d38464191bab7cb1d8040602c8f3b7207447ad0501cdbcc2e40ee84c682bb9L139

Best regards,
Mateusz

**Comment by mhagnumdw** [comment]
I just came across this issue having the same problem as you.

**Comment by mhagnumdw** [comment]
Given what was fixed in https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/225 , given that that fix ended up being undone and alerted here https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/577#pullrequestreview-1349778894 , and given that it is a real issue, then would we just have to open a PR to undo this https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/577#pullrequestreview-1349778894 ? Is that it folks ? Or is there a workaround without changing the plugin ?

**Comment by mc1arke** [comment]
@mhagnumdw if the plugin is doing the wrong thing then it needs a bug ticket raised; this ticket is about introducing a new feature. I'm then happy to review a PR against that ticket if someone was to propose one.

**Comment by mhagnumdw** [comment]
@mc1arke and others, if you can help with the thought process... maybe the plugin is correct and we actually need a solution for a specific, but real, scenario.

I went back and reread this entire issue, the Sonar documentation, and the GitLab documentation.

The Sonar Quality Gate can be based either on the new code only or on the entire codebase.

For branches and Merge Requests, the Quality Gate is applied only on the new code, which makes sense given the [Clean as You Code](https://docs.sonarqube.org/9.9/user-guide/clean-as-you-code/) methodology.

So the plugin is reporting the Quality Gate result for the Merge Request changes only, **which I believe is correct.**

Now let's consider a practical example where the Merge Request changes just a single line of code in a part of the codebase that doesn't currently have test coverage, and for whatever reason, it doesn't make sense to add tests for this line in this Merge Request. The total coverage of the whole codebase is 90%.

In this case, we’ll have the `SonarQube` job running in the dynamically created `external` stage, and it will fail. It fails because the Quality Gate failed - since the new code coverage is 0%.

On the GitLab side, this leads to two situations in the Merge Request:

* The pipeline is marked as failed, making it impossible to approve the Merge Request if you have this kind of blocking rule configured.
* The overall coverage reported in GitLab will appear as only 45% = (90 + 0)/2, which is clearly misleading, as GitLab simply averages the total coverage with the new code coverage.

Honestly, I’m not sure what the best solution would be... Maybe the plugin shouldn’t create the `SonarQube` job (in the `external` stage), since when the Quality Gate fails, a thread is already opened in the Merge Request. And if your GitLab configuration blocks Merge Request with unresolved threads, this thread would already be a blocking. You also have the option to configure your own Sonar job to fail or not fail the pipeline using `allow_failure: true`.

Another option is to keep everything as it is and manually override the Merge Request approval, assuming you’ve done the proper validations - but personally, I don’t think that’s ideal.
